### ISSUE TITLE
Auto refresh directories after file operations by external programs

### ISSUE THREAD

**Issue by inoshy** [kamiyaa/joshuto#92]
Hi, when some file operations take place by external programs, like when I download something from my browser, joshuto does not show downloaded files until I refresh it manually. Is there a way for joshuto to listen to file operation events and automatically refresh itself when needed? thanks :)

**Comment by kamiyaa** [comment]
Yep! This is something that I've thought about but haven't figured out the best way to do it yet.
Best way as in efficient (no unnecessary listening), yet functional.

Feel free to take this on if you're feeling adventureous.

Side note:
From my experience (in C), listening to directories might cause a problem on special directories like `/sys`, `/dev`, `/proc`, etc.

**Comment by inoshy** [comment]
Hi zhao! Thanks for your reply :)

I did some research and found this crate [notify-rs](https://github.com/notify-rs/notify) that implements unix `Inotify` linux api. but from the example [here](https://github.com/notify-rs/notify#usage), it seems like, they are polling every 2 seconds for event notifications? that is not effective right? because lots of the time it will be unnecessary polling?

sorry I might sound stupid I dont know much programming stuffs, pardon me for that!

**Comment by DLFW** [comment]
Hi, Christmas vacation are starting and I might find some time to pick this up.

On a first thought, I would try to approach this like this:
* Let each `JoshutoDirList` watch its directory. This will cause all “historic” directories to be watched, but for my understanding they are cached to avoid re-reading them from disk every time again. So, updating them is anyway necessary.
* If a change happens, the UI needs to be redrawn. From what I saw on a first glimpse, this happens on every event in the `AppContext`s event queue. So, I would consider to pass (and maybe keep) a reference to the `AppContext` to each `JoshutoDirList`. That does not feel very elegant but it would be necessary to enable the dir-lists to emit an event to the event-loop.
* Without further optimization, this solution would cause also inactive tabs to trigger a UI redraw. That might be ok.

Any opinions?

@inoshy: My interpretation from [the notify-rs docs](https://docs.rs/notify/latest/notify/#modules) is that notify-rs uses polling only on Windows. On Linux, it's event driven. :)

**Comment by kamiyaa** [comment]
Hey @DLFW thanks!

> - If a change happens, the UI needs to be redrawn. From what I saw on a first glimpse, this happens on every event in the AppContexts event queue. So, I would consider to pass (and maybe keep) a reference to the AppContext to each JoshutoDirList. That does not feel very elegant but it would be necessary to enable the dir-lists to emit an event to the event-loop.

I would avoid this. I'm not even sure if Rust will allow this because `AppContext` owns `History` which owns `JoshutoDirList`. If you add a reference to `AppContext` to `JoshutoDirList` it will cause some weird ownership problems. You can probably work around this but I would avoid it.

## The solution I had in mind was:
 - Spawning a thread to listen on the current folder whenever the user changes directories. This thread will listen to the following events:
  1. Filesystem change (triggered by `notify-rs` or some other lib). This will send an event to the main loop to reload the current directory
  2. Message from the main loop (directory changed). This will kill the thread and main loop `thread.join()`, then spawn a new one.

Some challenges this had was:
1. Simultaneously listening for filesystem changes AND main loop events at the same time. I want to avoid having to spawn too many thread as they are expensive. Listening to many filesystem changes can also be expensive. You can wrap both of them up into an `enum` like:
```rust
pub enum FilesystemListenerEvent {
    FilesystemModified,
    DirectoryChanged
}
```
But the main issue is writing the code to listen to both. It looks like `notify-rs` just runs a for loop.
A workaround for this could be on each user input, check if the filesystem has changed.

# Actually...

Briefly taking a look at `notify-rs` again, it looks like it already allows you to pass in a `Sender`: https://docs.rs/notify/latest/notify/trait.Watcher.html#tymethod.new

In that case, we can just have a global `watcher` in `run::run` and modify the directory it watches whenever we switch directories.

```rust

let mut old_cwd = ...;
watcher.watch(old_cwd)

while quit != ... {
    ...
    if (command is something that would change the cwd) {    // cd, parent_directory, open on directory, switch tabs, etc.
        watcher.unwatch(old_cwd);
        old_cwd = new_cwd;
        watcher.watch(old_cwd);
    }
}
```

**Comment by DLFW** [comment]
Hi! Thanks for all the thoughts on this!
I need to think about it again and take a look into the code again (really bed-time here...), but one doubt which popped up are the cached `JoshutoDirList`s in the `JoshutoHistory`s.

If we are in a dir A, then cd into B, have a file deleted externally in A, and then cd back to A, Joshuto would show the outdated content with the deleted file still in, wouldn't it? For my understanding, the history acts as a cache to avoid re-reading the dir.

One more thought: FS updates should be watched for all three panes. Having only one of them updated can be more confusing for a user, I guess.

**Comment by kamiyaa** [comment]
> I need to think about it again and take a look into the code again (really bed-time here...), but one doubt which popped up are the cached JoshutoDirLists in the JoshutoHistorys.

Yes `JoshutoHistory` is basically a wrapper around a HashMap that maps `PathBuf` to `JoshutoDirList`.
This is how `joshuto` keeps track of everything. Each tab have their own separate `JoshutoHistory` to prevent conflicts.

> If we are in a dir A, then cd into B, have a file deleted externally in A, and then cd back to A, Joshuto would show the outdated content with the deleted file still in, wouldn't it? For my understanding, the history acts as a cache to avoid re-reading the dir.

Yes, you are correct that history acts as a cache.

However, ***In this specific case*** where we delete a file externally in A, and cd back to A, `joshuto` will actually be able to detect the update and update automatically. I've added some code in the past to make `joshuto` smarter by checking the modified time when we switch directories:

https://github.com/kamiyaa/joshuto/blob/main/src/commands/parent_directory.rs#L24

However, this is the exception rather than the rule. I've mostly been trying to strike a balance between being smart and performance. Excessive mtime checks can have performance penalties on slower storage mediums like over the network, MTP, etc. Although these mtime checks can technically also be pushed off into its own thread to reduce the penalty. And then send a message back stating if the cache needs updating.

> One more thought: FS updates should be watched for all three panes. Having only one of them updated can be more confusing for a user, I guess.

Yeah thats a good idea, will just require some more coordination.

**Comment by DLFW** [comment]
That was a bit tricky. The “update by mtime” didn't work, but I found it was working in the past.
Seems 7d8b6dd broke this feature. 

I'm not sure if that was just an accident, but I provided #110 to revert this detail.

**Comment by DLFW** [comment]
The above PR now also contains changes to update the directories when the user opens a directory (before, it did just update when going one directory _up_) and to update the left most pane when switching a tab.

Updating the other panes on tab-switching and the "cd" command were already working properly.

I think, now the approach described above by @kamiyaa should work.  :)

Are there any other ways to change the shown cwd which might not be covered by a soft-update yet?
I have checked "cd", "open", "parent_directory", and tab-switching.
*May I have missed some scenario where we need a soft-reload to ensure that the directories are up-to-date after a user-interaction?*

**Comment by kamiyaa** [comment]
> That was a bit tricky. The “update by mtime” didn't work, but I found it was working in the past. Seems [7d8b6dd](https://github.com/kamiyaa/joshuto/commit/7d8b6ddd19c00661f999370f40a4c414183deb3b) broke this feature.
> 
> I'm not sure if that was just an accident, but I provided #110 to revert this detail.

Yeah, I think I removed it due to performance concerns

**Comment by kamiyaa** [comment]
> The above PR now also contains changes to update the directories when the user opens a directory (before, it did just update when going one directory _up_) and to update the left most pane when switching a tab.
> 
> Updating the other panes on tab-switching and the "cd" command were already working properly.
> 
> I think, now the approach described above by @kamiyaa should work. :)
> 
> Are there any other ways to change the shown cwd which might not be covered by a soft-update yet? I have checked "cd", "open", "parent_directory", and tab-switching. _May I have missed some scenario where we need a soft-reload to ensure that the directories are up-to-date after a user-interaction?_

I can't think of any off the top of my head. If there are additional scenarios, we can always update them in the future. :+1:

**Comment by DLFW** [comment]
Hi! Provided #112.
Not building for Mac though. Anyway, it seems to work on Linux without performance hassles. If somebody wants to test around a little, that would be great. I won't be able to work on this at least until end of the week.
